gammapy/maps/tests/test_wcs.py

```diff
@@ -121,3 +122,25 @@ def test_wcsgeom_contains(npix, binsz, coordsys, proj, skydir, axes):
     if not geom.is_allsky:
         coords = [0.0, 0.0] + [ax.center[0] for ax in geom.axes]
         assert_allclose(geom.contains(coords), np.zeros((1,), dtype=bool))
+
+def test_wcsgeom_solid_angle():
+    binsz=1.0
+    npix=100
+
+    # create dummy axes
+    axis1 = MapAxis.from_edges(np.arange(0,4))
+    axis2 = MapAxis.from_edges(np.arange(10,20))
+    # Test using cartesian geometry
+    geom = WcsGeom.create(skydir=(0, 0), npix=(npix, npix), binsz=binsz, coordsys='GAL',proj='CAR',axes=[axis1,axis2])
+
+    solid_array = geom.solid_angle()
+
+    # Check array size
+    assert solid_array.shape == (9,3,100,100)
```

Comment:
Maybe make the solid angle array 2D to avoid wasting memory by duplicating the same values for the extra axes?
I would do that and then leave broadcasting against extra axes (if needed) to callers using the solid angle arrays.

I think there is a method on geom to get the geom for the spatial part only, so should be easy to implement, no?

Reply:
I think this could be implemented either way, i.e. only return 2D or ND or add a keyword argument so that the caller can get either version.

But in any case, mentioning `geom.to_image().solid_angle()` in the docstring here would be appropriate, as well as how to work with the memory-efficient way, i.e. get a 2D solid angle array, and then multiply it against an ND map using numpy broadcasting or some other way to get multiplication of 2D array with ND map to work, i.e. what one needs for background, model evaluation, ... to multiply with this solid angle array.

Can you please decide on the implementation, but add a few lines to the docstring to explain how to work with this?

Reply:
A possibility could be to pass an option to geom.solid_angle(mode='full') to have the full ND array and mode='2D' to have only the 2D array.
Does it sound better like this?

Reply:
I was going to suggest to always return 2D and only explain in the docstring how to multiply that with an ND map (or if really needed, convert it to ND with one or two lines of Numpy/ Maps code).

But now I realise that I'm forgetting about the case of variable map pixel `npix` size.
I think those multi-resolution maps already exist, see the last example in this section:
http://docs.gammapy.org/dev/maps/index.html#constructing-with-factory-methods
and they require an ND (or at least > 2D) solid angle array.

Does that case work in your current implementation?
If yes, can you add an extra test case to show that it does?
E.g. two bins that differ by factor 2, and sum of solid angles in each plane should be almost the same, or assert just on one pixel from each plane, showing that it's roughly factor 4 different in solid angle.

@registerrier - I realise that my comments are always towards extending implementation and tests, but really we're better served by small and quick pull requests to have the basic 3D working, and then to extend later. So please finish up the PR as you think best, possibly leaving TODO or other comments for us with ideas what to improve at a later time.